src/ucs/sys/event_set.c

```diff
@@ -134,36 +135,58 @@ ucs_status_t ucs_event_set_del(ucs_sys_event_set_t *event_set, int event_fd)
     return UCS_OK;
 }
 
-ucs_status_t ucs_event_set_wait(ucs_sys_event_set_t *event_set, int timeout_ms,
-                                ucs_event_set_handler_t event_set_handler,
-                                void *arg)
+unsigned ucs_event_set_wait(ucs_sys_event_set_t *event_set, int max_events,
+                            int timeout_ms,
+                            ucs_event_set_handler_t event_set_handler,
+                            void *arg, ucs_status_t *status_p)
 {
-    int nready;
+    int read_events = 0;
     struct epoll_event ep_events[UCS_EVENT_EPOLL_MAX_EVENTS];
-    int i;
-
-    if (event_set_handler == NULL) {
-        return UCS_ERR_INVALID_PARAM;
-    }
-
-    nready = epoll_wait(event_set->epfd, ep_events, UCS_EVENT_EPOLL_MAX_EVENTS,
-                        timeout_ms);
-    if ((nready < 0) && (errno != EINTR)) {
-        ucs_error("epoll_wait() failed: %m");
-        return UCS_ERR_IO_ERROR;
-    }
-    ucs_trace_data("epoll_wait(epfd=%d, timeout=%d) returned %d",
-                   event_set->epfd, timeout_ms, nready);
-
-    for (i=0; i < nready; i++) {
-        int events = ucs_event_set_map_to_events(ep_events[i].events);
-        event_set_handler(ep_events[i].data.fd, events, arg);
-    }
-    return UCS_OK;
+    int i, max_wait_events, nready, events;
+
+    ucs_assert(event_set_handler != NULL);
+
+    do {
+        max_wait_events = ucs_min(max_events - read_events,
+                                  UCS_EVENT_EPOLL_MAX_EVENTS);
+        nready = epoll_wait(event_set->epfd, ep_events,
+                            max_wait_events, timeout_ms);
+        if (nready < 0) {
+            if (errno == EINTR) {
+                /* TODO: here we should take care about timeout_ms
+                 * (i.e. decrease it) but it is ok to not do this */
+                continue;
+            }
+            *status_p = UCS_ERR_IO_ERROR;
+            return 0;
+        }
+        ucs_trace_data("epoll_wait(epfd=%d, timeout=%d) returned %d",
+                       event_set->epfd, timeout_ms, nready);
+
+        for (i = 0; i < nready; i++) {
+            events = ucs_event_set_map_to_events(ep_events[i].events);
+            event_set_handler(ep_events[i].data.ptr, events, arg);
+        }
+
+        read_events += nready;
+    } while ((read_events < max_events) && (nready == max_wait_events));
```

Comment:
why need this cycle? Can we go with just one successful iteration?

Reply:
Hello @hiroyuki-sato 

We discussed this with @brminich and decided that's would be better to manage responsibilities between layers in a different way.

This loop will affect UCS/ASYNC/THREAD, so the loop should be done in TCP
here we should take care about timeout that is needed for UCS/ASYNC

So, we propose the following change:
```
ucs_status_t ucs_event_set_wait(ucs_sys_event_set_t *event_set,
                                unsigned max_events, int timeout_ms,
                                ucs_event_set_handler_t event_set_handler,
                                void *arg, unsigned *read_events)
{
    struct epoll_event ep_events[UCS_EVENT_EPOLL_MAX_EVENTS];
    unsigned i, max_wait_events;
    int nready, events;

    ucs_assert(event_set_handler != NULL);

    max_wait_events = ucs_min(max_events, UCS_EVENT_EPOLL_MAX_EVENTS);
    nready = epoll_wait(event_set->epfd, ep_events, max_wait_events, timeout_ms);
    if (nready < 0) {
        *read_events = 0;

        if (errno == EINTR) {
            return UCS_INPROGRESS;
        }

        ucs_fatal("epoll_wait() failed: %m");
        return UCS_ERR_IO_ERROR;
    }
    
    ucs_trace_data("epoll_wait(epfd=%d, timeout=%d) returned %d",
                   event_set->epfd, timeout_ms, nready);

    for (i = 0; i < nready; i++) {
        events = ucs_event_set_map_to_events(ep_events[i].events);
        event_set_handler(ep_events[i].data.ptr, events, arg);
    }

    *read_events = nready;
    if  ((nready < max_events) && (nready == UCS_EVENT_EPOLL_MAX_EVENTS)) {
        return UCS_INPROGRESS;
    }

    return UCS_OK;
}
```
```
/**
 * Wait for an I/O events
 *
 * @param [in]  event_set          Event set created by ucs_event_set_create.
 * @param [in]  max_events         Maximum wait events.
 * @param [in]  timeout_ms         Timeout period in ms.
 * @param [in]  event_set_handler  Callback functions.
 * @param [in]  arg                User data variables.
 * @param [out] read_events        number of read events.
 *
 * @return return UCS_OK on success, UCS_INPROGRESS - call was interrupted by a
 *         signal handler or there are probably more events to read,
 *         UCS_ERR_IO_ERROR - an error occurred.
 */
ucs_status_t ucs_event_set_wait(ucs_sys_event_set_t *event_set,
                                unsigned max_events, int timeout_ms,
                                ucs_event_set_handler_t event_set_handler,
                                void *arg, unsigned *read_events);
```

Reply:
I modified ucs_event_set_wait with your proposed way.